presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.resourceGroups;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+final class FairSharingQueue<E>
+        implements Queue<E>
+{
+    private final Map<E, Node<E>> index = new LinkedHashMap<>();
+
+    public boolean addOrUpdate(E element, Usage usage)
+    {
+        Node<E> node = index.get(element);
+        if (node != null) {
+            node.update(usage);
+            return false;
+        }
+
+        node = new Node<>(element, usage);
+        index.put(element, node);
+        return true;
+    }
+
+    @Override
+    public E peek()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean contains(E element)
+    {
+        return index.containsKey(element);
+    }
+
+    @Override
+    public boolean remove(E element)
+    {
+        Node<E> node = index.remove(element);
+        return node != null;
+    }
+
+    @Override
+    public E poll()
+    {
+        Collection<Node<E>> candidates = index.values();
+        long totalShare = 0;
+        long totalUtilization = 1; // prevent / by zero
+
+        for (Node<E> candidate : candidates) {
+            totalShare += candidate.getShare();
+            totalUtilization += candidate.getUtilization();
+        }
+
+        List<Node<E>> winners = new ArrayList<>();
+        double winnerFactor = 1;
+
+        for (Node<E> candidate : candidates) {
+            double share = 1.0 * candidate.getUtilization() / totalUtilization;
```

Comment:
`share` -> `actualShare` or `actualUsage` or `actualSth`. Then you compare it against `expectedSth` and you calculate the `deficit`/`delta`/`diff`.